File: src/main/js/formulas/math/LN.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class LN extends AbstractFormula {
  static SELF: LN = new LN();
  NAME = FormulaName.LN;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (value <= 0) {
      return new NumException(`LN parameter 1 value is ${value}. It should be greater than 0.`);
    }
    return Math.log(value);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/LN.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NumException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class LN extends AbstractFormula {
  public static FormulaName NAME = FormulaName.LN;
  public static LN SELF = new LN();

  public LN() {
    super();
  }

  public LN(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    Double value = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0])));
    if (value <= 0) {
      return new NumException(String.format("LN parameter 1 value is %f. It should be greater than 0.", values));
    }
    return Math.log(value);
  }
}
